---
title: "Reporting and analysis of BPS calls, without references"
---

```{r,read_in_datar_with_base_script,cache=T}
#rmarkdown::render("bps_calls.Rmd",output_dir=getwd(),knit_root_dir=getwd())

message("Reading in helper script for functions")
source("drcigar.R")
```

```{r,libs,cache=T}
library(tidyverse)
library(magrittr)
library(viridis)
library(parallel)
library(ggrepel)
library(future)
library(future.apply)
```

```{r,cache=T}
args <- commandArgs(trailingOnly=TRUE)
try(
tryCatch(
    plan(cluster,workers=args[1])
    ,
    plan(cluster,workers=32)
    )
)
```

```{r,some more funcs,cache=T}
delimiter <- function(x='raw') {
    if (x=='raw') return('---')
    else if (x=='se') return('\\-\\-\\-')
    else if (x=='de') return('\\\\-\\\\-\\\\-')
}

#parse_sam_flags
#cs_parser
#summarize_cs
#cigar_string_parser 
#summarize_cigar
#md_parser
#summarize_md

parse_cleaned_sam <- function(filename) {
    # You may wonder, why is this fella doing a round-trip read and write of the
    # SAM file before reading it in? Well, there's a weird bug in the `readr`
    # package that chokes on certain encoding on SAM files, it's like a byte
    # position in the file thing. Super weird. Reported on github, but they
    # ain't interested in fixing it, so this dirty hack works...
#    writeLines(
#        readLines(pipe(paste0("cat '",filename,"' | grep -v '^@'"))) 
#        ,paste0(filename,".clean"))
   # Then we read it back in
# 
#    read_tsv(
#            paste0(filename,".clean"), 
#            col_names=paste0("X",seq(1,100)),
#            comment="@"
#,n_max=1e3
#            ) %>%
       # Everything after 11th column should be tags
#        unite(col='tags',c(matches('X[1][2-9]'),matches('X2[0-9]')),
#            sep='_',na.rm=T) %>%
# UPDATE now we skip the above, not because the above problem, but because 
# it seems that the columns only stop at 12, which is perplexing, so instead
# readlines and split manually. So found two bugs in readr, I'm sure the first
# is in readr, the second I am pretty sure it's reading in the '\t*' in a SAM
# file without qualities to skip the rest, or something. Very strange
    returner <- readLines(pipe(paste0("cat '",filename,
            "' | grep -v '^@' "))) %>%
        strsplit(x=.,split="\t") %>%
        future_lapply(function(x){
                x[12] <- paste(x[12:length(x)],collapse="\t"); 
                names(x) <- paste0("X",seq(1,12));
                return(x[1:12])
            }) %>%
        bind_rows() %>%
        mutate(flag=future_lapply(X2,parse_sam_flags)) %>% select(-X2) %>%
        rename(ref=X3,seq=X10,qual=X11,cigar=X6,tags=X12) %>%
        mutate(tags=future_lapply(tags,function(x){
                str_split_1(x,pattern='\t') %>%
                    { setNames(object=., nm=toupper(substring(.,1,5)) ) } %>% return
            })) %>%
        mutate(nm=unlist(future_lapply(tags,function(x){
                    as.numeric(str_match(unlist(x)['NM:I:'],"NM:i:([^\t]+)")[2])
                }))) %>%
        mutate(md=future_lapply(tags,function(x){
                    md_string <- str_match(unlist(x)['MD:Z:'],"MD:Z:([^\t]+)")[2]
                    if(is.na(md_string)) {
                        return(tibble())
                    } else { 
                        return(summarize_md(md_parser(md_string)))
                    }
                })) %>%
        mutate(cs=future_lapply(tags,function(x){
                    cs_string <- str_match(unlist(x)['CS:Z:'],"[Cc][Ss]:[Zz]:(.+)")[2]
                    if(is.na(cs_string)) {
                        return(tibble())
                    } else { 
                        return(summarize_cs(cs_parser(cs_string)))
                    }
                })) %>%
        mutate(cigar_stats=future_lapply(cigar,function(x){
                        return(summarize_cigar(cigar_string_parser(x)) )
                })) 
        # And return it
    return(returner)
}
#tmp <- tibble(filenames=#list.files(path="./","*_to_ref.sam$")) %>%
#    head(1) %>%
#    mutate(
#        plasmid=sub("_payloads.sam","",filenames),
#        sam=map(filenames,parse_cleaned_sam)
#        )
#tmp$sam[[1]] %>%
#    filter(unlist(map(flag,
#                function(x){
#T
#                    #!length(intersect(x,c('supp_aln','secondary_aln')))
#                }))) %>% head() %>% data.frame
##payloads_datar %T>% {print(dim(.))} %>% sample_n(10)
#tmp <- parse_cleaned_sam("ref_align_001/428-lasso---r104_e81_sup_g610---1064-937-lasso---cutoutdonor---_to_ref_target.sam")
#tmp 

```


```{r,start and read stuff in, cache=T}

this_id <- readLines("id")
assembly_method <- readLines("assembly_method")
readlenclust <- NULL

if (assembly_method == '[msa]') {

    assemblies <- tibble( 
            filenames=Sys.glob("fasta_per_well_*/*/*fasta")
            ) %>%
        mutate(name=basename(filenames)) %>%
        separate(name,into=c('sample','code','position','misc',
                                'medaka','plasmid','postass'),
            sep=delimiter(),remove=T) %>%
        mutate(postass=ifelse(is.na(postass),misc,postass)) %>%
        mutate(postass=str_remove(postass,"\\.fasta")) %>%
        select(-misc) %>%
        mutate(
            file=future_lapply(filenames,function(x){
                    bind_rows(map(readLines(pipe(paste0("cat '",x,"' ",
                                "| sed 's/^>\\(.*\\)$/>\\1/' ",
                                "| tr -d '\\n'",
                                "| tr '>' '\\n'",
                                "| grep -v '^$' ",
                                "| tr '' '\t'"))) 
                        ,function(y){
                            setNames(unlist(strsplit(y,split="\\t")),c('name','seq'))
                        }))
                    })
            ) %>%
        unnest(file) %>% select(-filenames) %>% 
        mutate(name=str_remove(name,"\\s.*$")) %>%
        separate(name,into=c('nsample','ncode','nposition','misc'),
            sep=delimiter(),remove=F) %>% 
        select(-c('nsample','ncode','nposition'))
    assemblies %>% head

} else if (assembly_method == '[flye]') {

    assemblies <- tibble( 
            filenames=Sys.glob("fasta_per_well_*/*/*fasta")
            ) %>%
        mutate(name=basename(filenames)) %>%
        # removed contig parsing (misc) because trycycler doesn't seem to do 
        # separate contigs? I probably parsed it wrong, so maybe watch out if 
        # there's multiple assemblies per well
        #separate(name,into=c('sample','code','position','cluster','misc',
        separate(name,into=c('sample','code','position','cluster',
                                'medaka','plasmid','postass'),
            sep=delimiter(),remove=T) %>%
        mutate(misc=NA) %>%
        mutate(postass=ifelse(is.na(postass),misc,postass)) %>%
        mutate(postass=str_remove(postass,"\\.fasta")) %>%
        select(-misc) %>%
        mutate(
            file=future_lapply(filenames,function(x){
                    bind_rows(map(readLines(pipe(paste0("cat '",x,"' ",
                                "| sed 's/^>\\(.*\\)$/>\\1/' ",
                                "| tr -d '\\n'",
                                "| tr '>' '\\n'",
                                "| grep -v '^$' ",
                                "| tr '' '\t'"))) 
                        ,function(y){
                            setNames(unlist(strsplit(y,split="\\t")),c('name','seq'))
                        }))
                    })
            ) %>%
        unnest(file) %>% select(-filenames) %>% 
        mutate(name=str_remove(name,"\\s.*$")) %>%
        separate(name,into=c('nsample','ncode','nposition','cluster2','misc'),
            sep=delimiter(),remove=F) %>% 
        select(-c('nsample','ncode','nposition','cluster2'))
    assemblies %>% head

    readlenclust <- tibble( 
            filenames=Sys.glob("readlen_per_cluster_*/*tsv")
            ) %>%
        mutate(name=basename(filenames)) %>%
        separate(name,into=c('experiment','medaka','plasmid'),
            sep=delimiter(),remove=T) %>%
        mutate(
            file=future_lapply(filenames,function(x){
                    bind_rows(map(readLines(x) 
                        ,function(y){
                            z <- unlist(strsplit(y,split="\\t"))
                            name_is <- str_remove(basename(z[1]),".fastq")
                            return(
                                tibble(name_is,
                                    lengths=list(as.numeric(z[2:length(z)]))
                                    ))
                        }))
                    })
            ) %>%
        unnest(file) %>% select(-filenames) %>%
        separate(name_is,into=c('sample','code','position','cluster'),
            sep=delimiter(),remove=T) 
    readlenclust

}

```

```{r,read in target aln,cache=T}

aln_to_target <- tibble()


if (assembly_method == '[msa]') {

    try( {
    aln_to_target <- tibble( 
            filenames=Sys.glob("ref_align_*/*_to_ref_target.sam")
            ) %>%
        mutate(
            basename=sub(paste0(delimiter(),"_.*?$"),"",basename(filenames)),
            sam=map(filenames,parse_cleaned_sam)
            ) %>%
        separate(basename,into=c('experiment','medaka','plasmid','postass'),
            sep=delimiter()) %>%
        unnest(sam) %>%
        separate(X1,into=c("sample",'code','position','misc'),
            sep=delimiter(),extra="merge")  %>% 
        select(-qual) %>%
        filter(unlist(future_lapply(flag,
                    function(x){
                        !length(intersect(x,c('supp_aln','secondary_aln')))
                    }))) %>%
        mutate(length=nchar(seq)) %>%
        mutate(md=future_lapply(tags,function(x){
                    md_string <- str_match(x['MD:Z:'],"MD:Z:([^\t]+)\t?")[2]
                    if(is.na(md_string)) {
                        return(tibble())
                    } else { 
                        return(summarize_md(md_parser(md_string)))
                    }
                })) %>%
        unnest(md) %>%
        unnest(cigar_stats) %>%
        select(
            experiment,plasmid,medaka,sample,code,position,ref,
            postass,nm,length,cigar,misc,
            md_number_of_deletion,
            md_length_of_deletion,
            md_number_of_change,
            md_length_of_change,
            md_total_query_length,
            md_total_ref_length,
            cigar_number_of_insertion,
            cigar_length_of_insertion,
            cigar_number_of_deletion,
            cigar_length_of_deletion,
            cigar_number_of_substitution,
            cigar_length_of_substitution,
            cigar_number_of_soft_clip,
            cigar_length_of_soft_clip,
            seq
            ) 
    #aln_to_target %T>% {print(dim(.))} %>% select(-seq) %>% sample_n(10) %>% data.frame
    aln_to_target %T>% {print(dim(.))} %>% select(-seq) %>% sample_n(10) %>% data.frame
    })

} else if (assembly_method == '[flye]') {

    try( {
    aln_to_target <- tibble( 
            filenames=Sys.glob("ref_align_*/*_to_ref_target.sam")
            ) %>%
        mutate(
            basename=sub(paste0(delimiter(),"_.*?$"),"",basename(filenames)),
            sam=map(filenames,parse_cleaned_sam)
            ) %>%
        separate(basename,into=c('experiment','medaka','plasmid','postass'),
            sep=delimiter()) %>%
        unnest(sam) %>%
        separate(X1,into=c("sample",'code','position','cluster','misc'),
            sep=delimiter(),extra="merge")  %>% 
        select(-qual) %>%
        filter(unlist(future_lapply(flag,
                    function(x){
                        !length(intersect(x,c('supp_aln','secondary_aln')))
                    }))) %>%
        mutate(length=nchar(seq)) %>%
        mutate(md=future_lapply(tags,function(x){
                    md_string <- str_match(x['MD:Z:'],"MD:Z:([^\t]+)\t?")[2]
                    if(is.na(md_string)) {
                        return(tibble())
                    } else { 
                        return(summarize_md(md_parser(md_string)))
                    }
                })) %>%
        unnest(md) %>%
        unnest(cigar_stats) %>%
        select(
            experiment,plasmid,medaka,sample,code,position,ref,
            postass,nm,length,cigar,cluster,misc,
            md_number_of_deletion,
            md_length_of_deletion,
            md_number_of_change,
            md_length_of_change,
            md_total_query_length,
            md_total_ref_length,
            cigar_number_of_insertion,
            cigar_length_of_insertion,
            cigar_number_of_deletion,
            cigar_length_of_deletion,
            cigar_number_of_substitution,
            cigar_length_of_substitution,
            cigar_number_of_soft_clip,
            cigar_length_of_soft_clip,
            seq
            ) 
    #aln_to_target %T>% {print(dim(.))} %>% select(-seq) %>% sample_n(10) %>% data.frame
    aln_to_target %T>% {print(dim(.))} %>% select(-seq) %>% sample_n(10) %>% data.frame
    })

}

```


```{r,and the purities,cache=T}

if (assembly_method == '[msa]') {
    aln_to_polished <- tibble( 
            filenames=Sys.glob("purity_aln_*/*_pairwise_aln.tsv")
            ) %>%
        mutate(
            basename=sub(paste0(delimiter(),"_.*?$"),"",filenames),
            content=future_lapply(filenames,function(x){
                        read_tsv(pipe(paste0("cat '",x,"'")), 
                            col_names=c('id','ref','aln_score',
                                    'length','payload_seq','ref_seq')
                            )
                    })
            ) %>%
        separate(basename,into=c('experiment','medaka','plasmid'),sep=delimiter()) %>%
        mutate(experiment=basename(experiment)) %>%
        unnest(content) %>%
        separate(ref,into=c('sample','code','position','misc'),
                sep=delimiter(),extra="merge") %>%
        select(-filenames) %>%
        mutate(aln_frac=aln_score/length) 
        #group_by(ref,plasmid) %>%
    aln_to_polished %T>% {print(dim(.))} %>% sample_n(10) %>% data.frame
} else if (assembly_method == '[flye]') {

    aln_to_polished <- tibble( 
            filenames=Sys.glob("purity_aln_*/*_oriented_to_ref.tsv")
            ) %>%
        mutate(
            basename=sub(paste0(delimiter(),"_.*?$"),"",filenames),
            content=future_lapply(filenames,function(x){
                        read_tsv(pipe(paste0("cat '",x,"'")), 
                            col_names=c('id','ref','matches',
                                    'query_length','ref_seq')
                            )
                    })
            ) %>%
        separate(basename,into=c('experiment','medaka','plasmid'),sep=delimiter()) %>%
        mutate(experiment=basename(experiment)) %>%
        unnest(content) %>%
        separate(ref,into=c('sample','code','position','misc'),
                sep=delimiter(),extra="merge",remove=F) %>%
        select(-filenames) %>%
        mutate(aln_frac=matches/query_length) 
        #group_by(ref,plasmid) %>%
    aln_to_polished %T>% {print(dim(.))} %>% sample_n(10) %>% data.frame

}
#names(aln_to_polished)
#    "experiment"     "plasmid"        "medaka" "sample"  "code"  "position" 
#   "ref"  "id"    "nm"  "cigar"     "seq"            

```

```{r,paf part not using it so,eval=F,echo=F}

parse_cleaned_paf <- function(filename) {
    print(paste0("writing out ",filename))
    writeLines(
        readLines(pipe(paste0("cat '",filename,"' | grep -v '^@'")))
        , paste0(filename,".clean"))
    print(paste0("reading back in ",filename))
    read_tsv(
            paste0(filename,".clean"), 
            col_names=paste0("X",seq(1,100))
            ) %>%
        unite('tags',c(matches('X[1][3-9]'),matches('X2[0-9]'))) %>%
        mutate(type=unlist(future_lapply(tags,
                    function(x){str_match(x,"tp:A:([PSIi]*)")[2]}))) %>%
        mutate(nm=unlist(future_lapply(tags,function(x){
                    as.numeric(str_match(x,"NM:i:([^_]+)_")[2])}))) %>%
        mutate(cs=future_lapply(tags,function(x){
                    cs_string <- str_match(x,"cs:Z:([^_]+)")[2]
                    if(is.na(cs_string)) {
                        return(tibble())
                    } else { 
                        return(summarize_cs(cs_parser(cs_string)))
                    }
                })) %>%
        mutate(cigar_stats=future_lapply(tags,function(x){
                    cigar_string <- str_match(x,"cg:Z:([^_]+)_")[2]
                    if(is.na(cigar_string)) {
                        return(tibble())
                    } else { 
                        return(summarize_cigar(cigar_string_parser(cigar_string)) )
                    }
                })) %>%
        rename(
            qlength=X2,
            qstart=X3,
            qend=X4,
            tlength=X7,
            tstart=X8,
            tend=X9,
            same_strand=X5,
            target=X6,
            matches=X10,
            aln_length=X11,
            ) %>%
        mutate(
            identity=matches/aln_length,
            ) %>%
        select(-tags,-X12) %>%
        unnest_wider(cigar_stats) %>%
        unnest_wider(cs) %>%
        return()
###    Col Type    Description
###    1   string  Query sequence name
###    2   int Query sequence length
###    3   int Query start coordinate (0-based)
###    4   int Query end coordinate (0-based)
###    5   char    ‘+’ if query/target on the same strand; ‘-’ if opposite
###    6   string  Target sequence name
###    7   int Target sequence length
###    8   int Target start coordinate on the original strand
###    9   int Target end coordinate on the original strand
###    10  int Number of matching bases in the mapping
###    11  int Number bases, including gaps, in the mapping
###    12  int Mapping quality (0-255 with 255 for missing)
###When alignment is available, column 11 gives the total number of sequence matches, mismatches and gaps in the alignment; column 10 divided by column 11 gives the BLAST-like alignment identity. When alignment is unavailable, these two columns are approximate. PAF may optionally have additional fields in the SAM-like typed key-value format. Minimap2 may output the following tags:
###    Tag Type    Description
###    tp  A   Type of aln: P/primary, S/secondary and I,i/inversion
###    cm  i   Number of minimizers on the chain
###    s1  i   Chaining score
###    s2  i   Chaining score of the best secondary chain
###    NM  i   Total number of mismatches and gaps in the alignment
###    MD  Z   To generate the ref sequence in the alignment
###    AS  i   DP alignment score
###    SA  Z   List of other supplementary alignments
###    ms  i   DP score of the max scoring segment in the alignment
###    nn  i   Number of ambiguous bases in the alignment
###    ts  A   Transcript strand (splice mode only)
###    cg  Z   CIGAR string (only in PAF)
###    cs  Z   Difference string
###    dv  f   Approximate per-base sequence divergence
###    de  f   Gap-compressed per-base sequence divergence
###    rl  i   Length of query regions harboring repetitive seeds
}
ass2ref_paf <- Sys.glob("purity_aln_*/*paf") %>%
    map(parse_cleaned_paf) %>% bind_rows() %>%
    separate(X1,into=c('id','sample','code','position'),
        sep=delimiter(),extra="merge") 
ass2ref_paf %T>% {print(dim(.))} %>% head(10) %>% data.frame

```



```{r,check what you got,cache=T}

assembly_method
assemblies %>% head
aln_to_target %>% head
aln_to_polished %>% head

```

```{r,calc purity,cache=T}

if (assembly_method=='[flye]') {
    #aln_to_polished %>% head %>% select(-seq,-cigar) %>% data.frame
    per_read_aln_frac <- aln_to_polished %>% 
        group_by(experiment,plasmid,medaka,sample,position,ref) 
} else if (assembly_method=='[msa]') {
    per_read_aln_frac <- aln_to_polished %>% 
        unite(col='ref',sep=delimiter(),remove=F,
            c(sample,code,position,misc)) %>%
        group_by(experiment,plasmid,medaka,sample,position,ref) %>%
        select(length,aln_frac)
}
per_read_aln_frac

per_read_aln_frac %>% pull(aln_frac) %>% hist(.,100)

g <- per_read_aln_frac %>%
    group_by(experiment,plasmid,medaka,sample,position,ref) %>%
    ggplot()+theme_bw()+
    aes(x=aln_frac)+
    stat_bin(geom='line',bins=100)+
    stat_bin(geom='point',bins=100)+
    scale_x_log10(limits=c(0.5,1),breaks=seq(0,1,0.05))+
    geom_vline(xintercept=c(0.9,0.95,0.97),col='red')+
    xlab("Fraction of each raw read that\ndoesn't match polished consensus assembly")+
    ylab("Count")+
    NULL
g

ggsave(paste0(Sys.Date(),"_alnpercs_per_well.jpeg"),g,width=7,height=7)

purity <- per_read_aln_frac %>%
    group_by(experiment,plasmid,medaka,sample,position,ref) %>%
    summarize(obs=length(aln_frac),purity=sum(aln_frac>0.9)/obs)
purity

threshold <- max(10,quantile(purity$obs,probs=.05))
g <- purity %>%
    ggplot()+theme_bw()+
    aes(x=obs)+
    stat_bin(geom='line',bins=50)+
    stat_bin(geom='point',bins=50)+
    scale_x_log10()+
    xlab("Number of raw reads per well,\nused to generate polished consensus assembly")+
    ylab("Count")+
    geom_vline(xintercept=threshold,col="red",linetype="dashed")+
    NULL
g

ggsave(paste0(Sys.Date(),"_counts_per_well.jpeg"),g,width=7,height=7)

g <- purity %>%
    ggplot()+theme_bw()+
    aes(x=obs,y=purity)+
    geom_point(alpha=0.2)+
    scale_y_log10()+
    coord_cartesian(ylim=c(0.5,1))+
    xlab("Number of raw reads per well")+
    ylab("Fraction of reads that are more than 90% identitcal")+
    #geom_vline(xintercept=threshold,col="red",linetype="dashed")+
    NULL
#g
#if ( ! "try-error" %in% is(try( g2 <- g+geom_density_2d() )) ) {
#    ggsave(paste0(Sys.Date(),"_counts_and_purity_per_well.jpeg"),g2,width=7,height=7)
#    g2
#} else {
    ggsave(paste0(Sys.Date(),"_counts_and_purity_per_well.jpeg"),g,width=7,height=7)
    g
#}

```

```{r,readlenclust,cache=T}

if (assembly_method=="[flye]") {

    rlstats <- readlenclust %>% 
        mutate(
            lstats=future_lapply(lengths,function(x){
                    tibble(mean_length=mean(x),sd_length=sd(x),
                            median_length=median(x),n_length=length(x)) %>%
                        mutate(cv_length=sd_length/mean_length) %>%
                        return()
                })
            ) %>%
        unnest(lstats)
    rlstats

    g <- rlstats %>% ggplot()+theme_bw()+
        aes(x=cv_length,y=median_length,col=n_length>100)+
        geom_point()
    g
    ggsave(paste0(Sys.Date(),"_cluster_size_median_length.jpeg"),g,width=7,height=7)

}

```



```{r,called,cache=T}

called <- assemblies %>%
    filter(position!='unknown') %>%
    { if("cluster"%in%colnames(.)){ . } else { mutate(.,cluster=NA) }; } %>%
    mutate(misc=ifelse(is.na(misc),"",misc)) %>%
    left_join(
        purity %>% rename(name=ref)
        ,by=c('sample','position','name','medaka','plasmid')
        ) %>%
    filter(!is.na(experiment)) %>%
    group_by(experiment,sample,position,medaka) %>%
    mutate(
        main_plasmid=plasmid[order(obs,decreasing=T)[1]],
        ) %>%
    mutate(
        pure=purity>0.90,
        length=unlist(map(seq,nchar))
        ) %>% 
    ungroup() %>%
    group_by(experiment,plasmid,sample,position) %>%
    select(experiment,sample,position,
        medaka,
        coverage=obs,length,
        main_plasmid,plasmid,
        sequence=seq,
        pure,code,cluster,misc,
        post_assembly=postass,
        starts_with("purity")
        ) %>% 
    arrange(experiment,sample,position) %>% 
#    distinct() %>%
    group_by(experiment,sample,position) %>%
    mutate(
#        right_length= length==med_length ,
#        pure = purity>0.90 & coverage>=cutoff_for_depth,
        total_coverage=sum(coverage),
        ) %>%
    ungroup() %>% 
    select(
        experiment,
        medaka,
        sample,
        position,
        coverage,
        post_assembly,
        pure,
        purity,
        main_plasmid,
        plasmid,
        cluster,
        misc,
        code,
        sequence,
        )
called

if (nrow(aln_to_target)) {
called_with_ref <- called %>%
    left_join( 
        aln_to_target %>%
            mutate(mistakes=md_length_of_deletion+md_length_of_change+
                cigar_length_of_insertion+cigar_length_of_soft_clip ) %>%
            { if("cluster"%in%colnames(.)){ . } else { mutate(.,cluster=NA) }; } %>%
            mutate(misc=ifelse(is.na(misc),"",misc)) %>%
            select(experiment,plasmid,sample,position,cluster,misc,
                medaka,
                post_assembly=postass,
                ref,mistakes,starts_with('cigar_'),starts_with('md_'),starts_with('cs_'))
        , by=c('experiment','medaka','plasmid','sample','position','cluster','misc','post_assembly'
            )
        )
called_with_ref
} else {
called_with_ref <- called %>%
    mutate( ref=NA, mistakes=NA, cigar_length_of_soft_clip=NA,
        md_number_of_change=NA, cigar_length_of_insertion=NA,
        cigar_length_of_deletion=NA, cigar_number_of_insertion=NA,
        cigar_number_of_deletion=NA
        )
called_with_ref
}

if (assembly_method=="[flye]") {
    called_ref_rlstats <- called_with_ref %>%
    left_join( 
        rlstats %>%
            filter(position!='unknown') %>%
            select(experiment,medaka,plasmid,sample,position,cluster,
                median_length,n_length,cv_length)
        , by=c('experiment','medaka','plasmid','sample','position','cluster')
        )
} else {
    called_ref_rlstats <- called_with_ref %>%
        mutate(median_length=NA,n_length=NA,cv_length=NA)
}




called_ref_rlstats %>% 
    select(experiment,medaka,sample,position,post_assembly,pure) %>%
    distinct() %>%
    group_by(experiment,pure) %>%
    tally()

called_ref_rlstats %>% 
    group_by(medaka,post_assembly,pure,correct=mistakes==0&!cigar_length_of_soft_clip) %>%
    tally()

called_ref_rlstats %>% 
    group_by(medaka,post_assembly,pure,correct_lt10mistakes=mistakes<10&!cigar_length_of_soft_clip) %>%
    tally()

called_ref_rlstats %>% 
    #filter(pure,mistakes<10&!cigar_length_of_soft_clip) %>%
    group_by(ref,plasmid,pure,mistakes<10,!cigar_length_of_soft_clip) %>% 
    tally()

called_ref_rlstats %>% 
    filter(pure,mistakes<10&!cigar_length_of_soft_clip) %>%
    group_by(ref) %>% 
    tally()

output <- called_ref_rlstats %>% 
    mutate(assembly_method=assembly_method) %>%
    rowwise() %>%
    mutate(length_sequence=nchar(sequence)) %>%
    select(
        experiment,
        medaka,
        plasmid,
        sample,
        position,
        positioning_code=code,
        coverage,
        well_purity=purity,
        assembly_method,
        assembly_detail=misc,
        post_assembly,
        length_sequence,
        cluster,
        cluster_n_reads=n_length,
        cluster_median_length=median_length,
        cluster_cv_length=cv_length,
        matches_target=ref,
        query_errors=mistakes,
        query_changes=md_number_of_change,
        query_change_length=md_length_of_change,
        query_insertions=cigar_number_of_insertion,
        query_insertion_length=cigar_length_of_insertion,
        query_deletions=cigar_number_of_deletion,
        query_deletion_length=cigar_length_of_deletion,
        query_clipped_length=cigar_length_of_soft_clip,
        sequence,
        ) %>%
    arrange(experiment,plasmid,sample,assembly_method,post_assembly,position,cluster) 
output

output %>%
    {names(.) <- str_c(names(.),seq_along(names(.)),sep='_');.} %>%
    write_tsv(file="calls.tsv")


```


